spark/src/main/java/org/apache/iceberg/spark/actions/BaseRewriteDataFilesSparkAction.java

```diff
@@ -149,7 +168,13 @@ public RewriteDataFiles filter(Expression expression) {
 
     try {
       Map<StructLike, List<FileScanTask>> filesByPartition = Streams.stream(fileScanTasks)
-          .collect(Collectors.groupingBy(task -> task.file().partition()));
+          .collect(Collectors.groupingBy(task -> {
+            if (task.file().specId() == table.spec().specId()) {
+              return task.file().partition();
+            } else {
+              return EmptyStruct.get();
```

Comment:
What was the failure? Is our hash or equals function not correct for empty structs? Should this use a `StructLikeMap` instead?

Reply:
The issue is for files with the old partitioning, if we have another way of making an empty struct here that's fine. I can't remember why I chose to make a new class here since it was a while ago now.

The core issue is
Say we have originally have a table which is partitioned Bucket(x, 5) meaning any original data files are written with values of x more or less randomly distributed in our data files. Then our table has a partitioning changed to something like Bucket(x, 10). Worst case scenario is that when we write we end up having to make 10 files for every partition in our original bucketing. This says let's just assume all those files created with old partitioning are best dealt with at the same time, rather than splitting them up using the old partitioning.

Reply:
Okay, so all data where we're changing the partitioning goes into a separate task. That makes sense to me. Probably worth a comment here to explain it.

You may also want to change how this works slightly and use a `StructLikeMap`. Otherwise, `EmptyStruct` looks a lot like structs for the unpartitioned spec. It would or wouldn't be equal depending on the order of the comparison, which probably isn't what you want.

Instead, I would keep a `StructLikeMap` for the current partitioning scheme and then create a separate group using a `List` for the rest.

Reply:
I'm not quite sure I understand what you are describing here as a fix.

Reply:
Actually I think I see now. Should have this in a bit

Reply:
`StructLike` makes no guarantees about `equals`/`hashCode` behavior, so using it as a map key is like using a `CharSequence` as a map key. Probably not a good idea because it will break if the underlying implementation changes or differs. I'd recommend using `StructLikeMap` that handles consistent behavior. But that requires using a specific struct type. Since you really only need the table spec's struct type, you can use that. Then keep any tasks that are not in the current table spec in a list, like this:

```java
StructLikeMap<FileScanTask> filesByPartition = StructLikeMap.create(table.spec().partitionType());
List<FileScanTask> tasksFromOtherSpecs = Lists.newArrayList();
Streams.stream(fileScanTasks).forEach(task -> {
    if (task.file().specId() != table.spec().specId()) {
      tasksFromOtherSpecs.add(task);
    } else {
      filesByPartition.put(task.file().partition(), task);
    }
  });
```

Reply:
The only issue I can think of here is that it doesn't fit well with the rest of the code which assumes all FileGroups have an associated partition value, I think empty is the right thing to pass since we are basically treating those files as unpartitioned.

Reply:
Trying something like
```
      StructLikeMap<List<FileScanTask>> filesByPartition = StructLikeMap.create(table.spec().partitionType());
      StructLike emptyStruct = GenericRecord.create(table.schema());

      fileScanTasks.forEach(task -> {
        /*
        If a task uses an incompatible partition spec the data inside could contain values which
        belong to multiple partitions in the current spec. Treating all such files as un-partitioned and
        grouping them together helps to minimize new files made.
        */
        StructLike taskPartition = task.file().specId() == table.spec().specId() ?
            task.file().partition() : emptyStruct;

        List<FileScanTask> files = filesByPartition.get(taskPartition);
        if (files == null) {
          files = Lists.newArrayList();
        }

        files.add(task);
        filesByPartition.put(taskPartition, files);
      });
      ```

Reply:
Yeah, you may need to pass an empty struct through the rest of the code. But it is probably a good idea not to use StructLike as a map key.